keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,775 @@
+        #   3. If any matchCondition evaluates to an error (but none are FALSE):
+        #      - If failurePolicy=Deny, reject the request
+        #      - If failurePolicy=NoOpinion, the error is ignored and the webhook is skipped
+      matchConditions:
+      # expression represents the expression which will be evaluated by CEL. Must evaluate to bool.
+      # CEL expressions have access to the contents of the SubjectAccessReview
+      # in the version specified by subjectAccessReviewVersion in the request variable.
+      # Documentation on CEL: https://kubernetes.io/docs/reference/using-api/cel/
+      #
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests to kube-system
+      - expression: request.resourceAttributes.namespace == 'kube-system'
+      # don't intercept requests from kube-system service accounts
+      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
+  - type: Node
+  - type: RBAC
+  - type: Webhook
+    webhook:
+      name: in-cluster-authorizer
+      authorizedTTL: 5m
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      failurePolicy: NoOpinion
+      connectionInfo:
+        type: InClusterConfig
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the
+config, but one authorizer each for other types. The ordering of this chain will
+be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
+`subjectAccessReviewVersion` accept values corresponding to flags
+`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
+`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
+respectively.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The user can define a CEL expression to determine whether a request needs to dispatched
+to the authz webhook for which the expression has been defined. The user would have access
+to a `request` variable containing a `SubjectAccessReview` object in the version specified
+by `subjectAccessReviewVersion`.
+
+The code path for enabling the above will only be triggered if the feature flag is enabled until 
+the feature flag is removed and this feature graduates to GA.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+Labels {along with possible values}:
+- `mode` {<authorizer_name>}
```

Comment:
> - `mode` {<authorizer_name>} when authorizer is a webhook, prepend `webhook_`

Reply:
There is a note a couple of lines below which specifies this information. I think that conveys the same information in addition to specifying what happens when there is only one Webhook and no name is specified.

Reply:
Can you pls explicitly add `when authorizer is a webhook, prepend webhook_` here so we do not forget it during implementation.